wsmaster/che-core-api-user/src/main/java/org/eclipse/che/api/user/server/UserManager.java

```diff
@@ -252,6 +253,34 @@ public long getTotalCount() throws ServerException {
      */
     public void remove(String id) throws ServerException, ConflictException {
         requireNonNull(id, "Required non-null id");
+        doRemove(id);
+        eventService.publish(new UserRemovedEvent(id));
+    }
+
+    @Transactional(rollbackOn = {RuntimeException.class, ServerException.class})
+    protected void doRemove(String id) throws ServerException {
+        UserImpl user;
+        try {
+            user = userDao.getById(id);
+        } catch (NotFoundException ignored) {
+            return;
+        }
+
+        eventService.publish(new BeforeUserRemovedEvent(user)).propagateException();
         userDao.remove(id);
     }
+
+    private class RemovePreferencesBeforeUserRemovedEventSubscriber extends CascadeEventSubscriber<BeforeUserRemovedEvent> {
+        @Override
+        public void onCascadeEvent(BeforeUserRemovedEvent event) throws Exception {
+            preferencesDao.remove(event.getUser().getId());
```

Comment:
nice catch, it makes sense to move this code from listeners to manager `doRemove`